### PR TITLE
implemented ->form_with()

### PR DESCRIPTION
**Pull request by fany**
Since my bank uses funny obfuscation techniques in their [online portal](https://www.sparkasse-nuernberg.de/) (changing names of and field names within forms upon each request), I needed a more flexible way to select forms.
I hope it is self-explanatory.

### CODE REVIEW

File: lib/WWW/Mechanize.pm [hunk 1]
```diff
@@ -1454,6 +1433,59 @@ sub form_with_fields {
...
+Returns undef if no form is found.
+
+
+=cut
+
+sub form_with {
+    my ( $self, %spec ) = @_;
+
+    my @forms = $self->forms or return;
+    while ( my ( $attr, $expected_value ) = each %spec ) {
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: `each` has some caveats, so I think I'd rather avoid it entirely.  Could we rework this bit?  http://blogs.perl.org/users/rurban/2014/04/do-not-use-each.html

File: t/local/form.t [hunk 2]
```diff
@@ -34,3 +34,8 @@ ok( !$mech->form_name('bargle-snark'), 'cannot select non-existent form' );
 my $form_id_pounder = $mech->form_id('pounder');
 isa_ok( $form_id_pounder, 'HTML::Form', 'Can select the form' );
 ok( !$mech->form_id('bargle-snark'), 'cannot select non-existent form' );
+
+my $form_with = $mech->form_with( class => 'test', id => undef );
+isa_ok( $form_with, 'HTML::Form', 'Can select the form without id' );
+is( $mech->current_form, $form_number_1,
+    'Form without id is now the current form' );
```
Comment (oalders) [inline t/local/form.t]: 👍  Thanks very much for including the test.

File: lib/WWW/Mechanize.pm [hunk 3]
```diff
@@ -1367,19 +1367,7 @@ Returns undef if no form is found.
...
-    my $nmatches = @matches;
-    if ( $nmatches > 0 ) {
-        if ( $nmatches > 1 ) {
-            $self->warn( "There are $nmatches forms named $form.  The first one was used." )
-        }
-        return $self->{current_form} = $matches[0];
-    }
-
-    return;
+    $self->form_with( name => $form );
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: Could we make this an explicit `return`?

File: lib/WWW/Mechanize.pm [hunk 4]
```diff
@@ -1398,18 +1386,9 @@ unless C<quiet> is enabled.
...
-            if @matches > 1;
-        return $self->{current_form} = $matches[0];
-    }
-    else {
-        $self->warn( qq{ There is no form with ID "$formid"} );
-        return undef;
-    }
+    defined( my $form = $self->form_with( id => $formid ) )
+      or $self->warn(qq{ There is no form with ID "$formid"});
+    $form;
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: Could we make this an explicit `return`?

File: lib/WWW/Mechanize.pm [hunk 5]
```diff
@@ -1454,6 +1433,59 @@ sub form_with_fields {
...
+    while ( my ( $attr, $expected_value ) = each %spec ) {
+        @forms = grep {
+            my $actual_value = $_->attr($attr);
+            if ( defined $expected_value ) {
+                defined $actual_value && $actual_value eq $expected_value;
+            }
+            else {
+                !defined $actual_value;
+            }
+        } @forms or return;
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: This multi-line grep, seems a bit complicated.  Can we make this a bit simpler?

File: lib/WWW/Mechanize.pm [hunk 6]
```diff
@@ -1454,6 +1433,59 @@ sub form_with_fields {
...
+                ? length $spec{$_}
+                      ? "$_ $spec{$_}"
+                      : "empty $_"
+                : "no $_",
+                sort keys %spec )
+          )
+          . '.  The first one was used.'
+    ) if @forms > 1;
+    $self->{current_form} = $forms[0];
+}
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: This is a bit hard to follow as well.  If we could simplify this or add some comments, that would be great.

**Comment by oalders** [comment]
Thanks very much for this.  I really like the idea.  I had a few comments about the implementation.  They're not showstoppers, but if you had time to address them, that would be great.

Also, feel free to add an appropriate comment to `Changes`.

**Comment by fany** [comment]
Thank you very much for your quick reaction and your comments!
I tried to address them and hope you like the reworked code.

File: lib/WWW/Mechanize.pm (hunk 5 above)
Comment (karenetheridge) [inline lib/WWW/Mechanize.pm]: That code simplifies down to:

        @forms = grep {
            !defined $expected_value || $_->attr($attr) eq $expected_value
        } @forms or return;

Comment (fany) [inline lib/WWW/Mechanize.pm]: Thanks for your suggestion, @karenetheridge!  However, this would IMHO impose two problems in case an attribute searched for is not present in the page:
* "Use of uninitialized value in string eq" warnings would be emitted.
* The case "attribute is present with empty value" would not be distinguished from "attribute is not present".

I have reworked this `grep` in the meantime, though:
https://github.com/libwww-perl/WWW-Mechanize/pull/48/commits/db3e94d96b4e4223d1b3359a326ed1713867224d
Maybe you like the new version.

Comment (oalders) [inline lib/WWW/Mechanize.pm]: The new version is a nice improvement.  This seems like the kind of place where https://metacpan.org/pod/List::Gather might make it a bit clearer what is going on, but I don't want you to rework this X times.  Could you add comments explaining your second point?

> The case "attribute is present with empty value" would not be distinguished from "attribute is not present".

I got that from the code, but it took me a while to work it out. ;)

**Comment by oalders** [comment]
Thanks for the lightning speed! 🚀 

Do you think we could add a test for forms which are not found?  It would be nice to get test coverage for the code path with the warning you added.

**Comment by fany** [comment]
OK, I added some tests and also factored out and commented the check for definedness and equality.
There is also an alternative version of `_equal()` suggested by my colleague @moritz, which I put in the [branch "moritz"](https://github.com/fany/WWW-Mechanize/tree/moritz).

File: lib/WWW/Mechanize.pm [hunk 7]
```diff
@@ -1456,13 +1456,7 @@ sub form_with {
...
     foreach my $attr ( keys %spec ) {
-        @forms = grep {
-            my $expected_value = $spec{$attr};
-            my $actual_value   = $_->attr($attr);
-            defined $expected_value
-              ? defined $actual_value && $actual_value eq $expected_value
-              : !defined $actual_value;
-        } @forms or return;
+        @forms = grep _equal( $spec{$attr}, $_->attr($attr) ), @forms or return;
     }
```
Comment (oalders) [inline lib/WWW/Mechanize.pm]: 👍

**Comment by oalders** [comment]
This looks good to me.  I'm happy with either implementation of `_equal()` so I'll leave that up to you and @moritz if you want to change it.  I think it's quite readable as it is, though.  @karenetheridge did you have any more comments?  I'm happy to merge and release. 

@fany thanks for your work and particularly for the iterations.

**Comment by oalders** [comment]
This has now been uploaded to CPAN.  😄
